syntax = "proto3";

package zookeeper;

enum OpType {
  UNKNOWN = 0;
  WRITE = 1;
  GUESS = 2;
  HEARTBEAT = 3;
  ELECTION = 4;
  ACK = 5;
  COMMIT = 6;
}

message Proposal {
  int32 epoch = 1;
  int32 seq = 2;
  OpType op = 3;
  string value = 4;
  string from_id = 5;
}

message Ack {
  int32 epoch = 1;
  int32 seq = 2;
  string from_id = 3;
  bool accepted = 4;
}

message Commit {
  int32 epoch = 1;
  int32 seq = 2;
  string value = 3;
}

message Heartbeat {
  string leader_id = 1;
  int64 timestamp = 2;
}

message Election {
  string candidate_id = 1;
  int32 epoch = 2;
  int32 seq = 3;
}

message Vote {
  string voter_id = 1;
  string vote_for = 2;
  int32 epoch = 3;
}

message Empty {}

service NodeService {
  rpc SendProposal(Proposal) returns (Ack);
  rpc SendCommit(Commit) returns (Empty);
  rpc SendHeartbeat(Heartbeat) returns (Empty);
  rpc StartElection(Election) returns (Vote);
  rpc KillSelf(Empty) returns (Empty);
}

message ClientRequest {
  OpType op = 1;
  string value = 2;
}

message ClientResponse {
  bool success = 1;
  string message = 2;
  string current_target = 3;
}

service ClientService {
  rpc SubmitRequest(ClientRequest) returns (ClientResponse);
}
